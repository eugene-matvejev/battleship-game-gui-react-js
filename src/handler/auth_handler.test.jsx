import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { AuthHandler } from './';

configure({ adapter: new Adapter() });

describe('<AuthHandler/>', () => {
    const props = {
        callback: () => { },
    }

    describe('render', () => {
        it('renders without crash', () => {
            shallow(<AuthHandler {...props} />);
        });
    });

    describe('inputs', () => {
        [{ type: 'password', field: 'password' }, { type: 'text', field: 'username' }].forEach(({ type, field }) => {
            it(`${field} field in state get altered with change`, () => {
                const component = shallow(<AuthHandler {...props} />);

                component.find(`input[type="${type}"]`).simulate('change', { target: { value: field } })

                expect(component.state()[field]).toBe(field);
            });
        });
    });

    describe('buttons', () => {
        it('if button.btn-submit been clicked handler will be called with expected payload', () => {
            const changedProps = Object.assign({}, props);
            let expectedPayload;

            changedProps.callback = (payload, onSuccess, onFail) => Promise
                .resolve(true)
                .then(() => { expectedPayload = payload; });

            const component = shallow(<AuthHandler {...changedProps} />);
            component.find('button.btn-submit').simulate('click');

            Promise
                .resolve(component)
                .then((c) => c.update())
                .then(() => expect(expectedPayload).toEqual({username: '', password: ''}));
        });
    });

    describe('async update', () => {
        [
            {success: true, desc: 'success'},
            {success: false, desc: 'fail'},
        ].forEach(({success, desc}) => {
            it(`on${desc[0].toUpperCase() + desc.slice(1)}`, () => {
                const changedProps = Object.assign({}, props);

                changedProps.callback = (payload, onSuccess, onFail) => Promise
                    .resolve(true)
                    .then(() => success ? onSuccess(payload) : onFail(payload))

                const component = shallow(<AuthHandler {...changedProps} />);
                component.find('button.btn-submit').simulate('click');

                Promise
                    .resolve(component)
                    .then((c) => c.update())
                    .then((c) => expect(c.state().color).toBe(success ? 'green' : 'red'));
            });
        });
    });
});
