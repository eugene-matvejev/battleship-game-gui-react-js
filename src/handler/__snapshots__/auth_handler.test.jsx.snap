// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthHandler/> render match snapshop 1`] = `
<div
  className="handler auth props.className"
>
  <h1>
    props.label
  </h1>
  <div
    className="input-wrapper"
  >
    <input
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="input-wrapper"
  >
    <input
      onChange={[Function]}
      type="password"
    />
  </div>
  <div>
    <button
      className="btn-submit"
      onClick={[Function]}
    >
      log in
    </button>
  </div>
  <Link
    replace={false}
    to="/sign-up"
  >
    sign up
  </Link>
  <Link
    replace={false}
    to="/password-reset"
  >
    forgotten password
  </Link>
</div>
`;
